File: lab_1/main.cpp
#include <iostream>

#include "byte_gen.h"

using namespace std;

// int main(){
//     // unsigned n = 4567;
//     // std::cout << n << std::endl;
//     // std::cout << ((n >> 1) | (n << 31)) << std::endl;
//     // std::cout << n + 2 << std::endl;

//     LongNumber p = LongNumber("cea42b987c44fa642d80ad9f51f10457690def10c83d0bc1bcee12fc3b6093e3");
//     LongNumber a = LongNumber("5b88c41246790891c095e2878880342e88c79974303bd0400b090fe38a688356");
//     LongNumber comp_p = LongDivMod(LongSub(p, LongNumber("1")), LongNumber("2"));
//     LongNumber T_0 = LongNumber("7d657c8f14f7389ad035a31fb6425b44a88470f4666b42ed20960d0ba4baf0db");
//     std::cout << LongCmp(comp_p, T_0) << std::endl;
// }

int main() {
    BuildInGenerator build_in_gen = BuildInGenerator();
    LehmerLowGenerator lehmer_low = LehmerLowGenerator(65537, 119, 4444444);
    LehmerHighGenerator lehmer_high = LehmerHighGenerator(65537, 119, 4444444);
    L20Generator l_20 = L20Generator(1);
    L89Generator l_89 = L89Generator(429496295);
    GeffeGenerator geffe = GeffeGenerator(34354672, 34656, 274563);
    WolframGenerator wolfram = WolframGenerator(32456);
    LibrarianGenerator baba_luda = LibrarianGenerator("Some string about warm love of Ron Uisli and his rat. Rat, rat, rat. How I like you my llittle liker of looking of sleeping ron.");
    BlumMikaliGenerator blum_mika = BlumMikaliGenerator("12323eac7823");
    BlumMikaliByteGenerator blum_mika_byte = BlumMikaliByteGenerator("12323eac7823");

    for(int i = 0; i < 100; ++i){
        for(auto b: blum_mika_byte.next()){
            std::cout << b << " ";
        }

        // std::cout << std::endl;
    }

    return 0;
}

File: lab_1/main.py
import io

import byte_gen
from byte_gen_test import signs_equality as se
from byte_gen_test import independence_of_signs as ios
from byte_gen_test import binary_sequence_uniformity as bsu
from byte_gen_test import chi_2_quantile as ch2
from my_timer import My_Timer


def bp(b):
    if b:
        return "True "
    
    return "False"

def main(rv_size=1000000):
    build_in_gen = byte_gen.BuildInGenerator();
    lehmer_low = byte_gen.LehmerLowGenerator(65537, 119, 4444444);
    lehmer_high = byte_gen.LehmerHighGenerator(65537, 119, 4444444);
    l_20 = byte_gen.L20Generator(1);
    l_89 = byte_gen.L89Generator(429496295);
    geffe = byte_gen.GeffeGenerator(34354672, 34656, 274563);
    wolfram = byte_gen.WolframGenerator(32456);
    
    trash_text = "Some string about warm love of Ron Uisli and his rat. Rat, rat, rat. How I like you my llittle liker of looking of sleeping ron."
    with io.open("lab_1/alice.txt", "r", encoding='utf-8') as f:
        trash_text = f.read()

    baba_luda = byte_gen.LibrarianGenerator(trash_text);
    blum_mika = byte_gen.BlumMikaliGenerator("12323eac7823");
    blum_mika_byte = byte_gen.BlumMikaliByteGenerator("12323eac7823");
    blum_blum_shuba = byte_gen.BlumBlumShubaGenerator("12323eac7823");
    blum_blum_shuba_byte = byte_gen.BlumBlumShubaByteGenerator("12323eac7823");

    gens = [
        build_in_gen,
        lehmer_low,
        lehmer_high,
        l_20,
        l_89,
        geffe,
        wolfram,
        baba_luda,
        blum_mika,
        blum_mika_byte,
        blum_blum_shuba,
        blum_blum_shuba_byte,
    ]

    timer_ = My_Timer()
    k = 1
    for gen in gens:
        print(f"\n{k:2d}) {gen.about()}")

        rv = [] 

        for i in range(0, rv_size):
            next = gen.next_byte()
            rv.append(next)
        
        print("alphas | equality | independence | bsu 10^1 | bsu 10^2 | bsu 10^3 | bsu 10^4")
        for a in [0.01, 0.05, 0.1]:
            print(f"{a:0.2f}   | {bp(se(rv, a))}    | {bp(ios(rv, a))}        | {bp(bsu(rv, a,10))}    | {bp(bsu(rv, a,100))}    | {bp(bsu(rv, a,1000))}    | {bp(bsu(rv, a,10000))}")

        print(f"Spend: {timer_.point(): 0.2f} s.")
        k += 1

    return 0;

if __name__ == "__main__":
    main(1000)

